core/src/main/java/org/elasticsearch/script/TemplateService.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.watcher.ResourceWatcherService;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+import static java.util.Collections.emptyMap;
+
+public class TemplateService implements ClusterStateListener {
+    public interface Backend extends ScriptEngineService {} // TODO customize this for templates
+
+    private static final Logger logger = ESLoggerFactory.getLogger(TemplateService.class);
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(logger);
+    private final Backend backend;
+    private final ScriptPermits scriptPermits;
+    private final CachingCompiler<String> compiler;
+
+    public TemplateService(Settings settings, Environment env,
+            ResourceWatcherService resourceWatcherService, Backend backend,
+            ScriptContextRegistry scriptContextRegistry, ScriptSettings scriptSettings,
+            ScriptMetrics scriptMetrics) throws IOException {
+        Objects.requireNonNull(scriptContextRegistry);
+
+        this.backend = backend;
+        this.scriptPermits = new ScriptPermits(settings, scriptSettings, scriptContextRegistry);
+        this.compiler = new CachingCompiler<String>(settings, env, resourceWatcherService,
```

Comment:
We hardcode `mustache` in so many places I figured we really only need to support a single template language at a time. If we want to support multiple then we can revisit, but for now I went with only one at a time.